optashift-employee-rostering-gwtui/src/main/java/org/optaplanner/openshift/employeerostering/gwtui/client/pages/rotation/ShiftBlob.java

```diff
@@ -0,0 +1,152 @@
+    private ShiftBlob twin;
+
+    ShiftBlob(final Shift shift,
+              final LocalDateTime baseDate,
+              final LinearScale<Long> scale) {
+
+        this.shift = shift;
+        this.scale = scale;
+        this.baseDate = baseDate;
+        this.sizeInGridPixels = getInitialSizeInGridPixels();
+        this.twin = getUpdatedTwin();
+    }
+
+    private ShiftBlob(final Shift shift,
+                      final LocalDateTime baseDate,
+                      final LinearScale<Long> scale,
+                      final ShiftBlob twin) {
+
+        this.shift = shift;
+        this.scale = scale;
+        this.baseDate = baseDate;
+        this.sizeInGridPixels = getInitialSizeInGridPixels();
+        this.twin = twin;
+    }
+
+    private long getInitialSizeInGridPixels() {
+        return scale.toGridPixels(minutesAfterBaseDate(shift.getTimeSlot().getEndDateTime())) - getPositionInGridPixels();
+    }
+
+    @Override
+    public Long getPositionInScaleUnits() {
+        return minutesAfterBaseDate(shift.getTimeSlot().getStartDateTime());
+    }
+
+    @Override
+    public void setPositionInScaleUnits(final Long positionInScaleUnits) {
+        shift.getTimeSlot().setStartDateTime(baseDate.plusMinutes(positionInScaleUnits));
+    }
+
+    @Override
+    public Long getSizeInGridPixels() {
+        return sizeInGridPixels;
+    }
+
+    @Override
+    public void setSizeInGridPixels(final Long sizeInGridPixels) {
+        this.sizeInGridPixels = sizeInGridPixels;
+        shift.getTimeSlot().setEndDateTime(shift.getTimeSlot().getStartDateTime().plusMinutes(scale.toScaleUnits(sizeInGridPixels)));
+    }
+
+    private long minutesAfterBaseDate(final LocalDateTime startDateTime) {
+        return Duration.between(baseDate.toInstant(UTC), startDateTime.toInstant(UTC)).getSeconds() / 60;
+    }
+
+    @Override
+    public ShiftBlob getUpdatedTwin() {
+
+        //FIXME: Maybe it's better to use the Viewport sizeInGridPixels instead of the end of the scale
+
+        final boolean hasAnyPartOffTheGrid =
+                getEndPositionInGridPixels() > scale.getEndInGridPixels() ||
+                        getPositionInGridPixels() < 0;
+
+        if (hasAnyPartOffTheGrid) {
+            final ShiftBlob twin = getTwin().orElseGet(this::newTwin);
+            final Long offset = (getPositionInGridPixels() < 0 ? 1 : -1) * scale.getEndInGridPixels();
+            twin.setPositionInScaleUnits(scale.toScaleUnits(getPositionInGridPixels() + offset));
+            twin.setSizeInGridPixels(sizeInGridPixels);
+            return twin;
+        } else {
+            return null;
+        }
+    }
+
+    public ShiftBlob newTwin() {
+
+        final TimeSlot timeSlot = new TimeSlot(
+                shift.getTimeSlot().getTenantId(),
+                shift.getTimeSlot().getStartDateTime(), //intentionally overwritten below
+                shift.getTimeSlot().getEndDateTime()); //intentionally overwritten below
```

Comment:
I don't see `shift.getTimeSlot().setStartDateTime(...)` or `timeSlot.setStartDateTime(...)` anywhere, so how is it overwritten?

Reply:
They're set in the `CircularBlobChangeHandler` class.